### ISSUE TITLE
Conflict between ContentNegotiation and Websockets

### ISSUE THREAD

**Issue by laurencegw** [ktorio/ktor#745]
### Ktor Version 
"1.0.0"

### Ktor Engine Used(client or server and name)
ktor-server-netty

### JVM Version, Operating System and Relevant Context
openjdk version "1.8.0_181"
ubuntu

### Feedback

At the moment it seems that `ContentNegotiation` plugin is incompatible with the `Websockets` plugin.
Possibly a feature/config is needed to restrict the routes that content negotiation occurs on?

When both are installed, any calls to the websocket endpoint just hangs.

#### Code gist - slightly modified from the websocket testing example.

    import io.ktor.application.install
    import io.ktor.gson.gson
    import io.ktor.http.cio.websocket.Frame
    import io.ktor.http.cio.websocket.readText
    import io.ktor.routing.routing
    import io.ktor.server.testing.withTestApplication
    import io.ktor.websocket.WebSockets
    import io.ktor.websocket.webSocket
    import kotlinx.coroutines.channels.ClosedReceiveChannelException
    import org.junit.jupiter.api.Assertions.assertEquals
    import org.junit.jupiter.api.Test
    
    class MyAppTest {
        @Test
        fun testConversation() {
            withTestApplication {
                application.install(WebSockets)
                application.install(io.ktor.features.ContentNegotiation) {
                    gson {}
                }
    
                val received = arrayListOf<String>()
                application.routing {
                    webSocket("/echo") {
                        try {
                            while (true) {
                                val text = (incoming.receive() as Frame.Text).readText()
                                received += text
                                outgoing.send(Frame.Text(text))
                            }
                        } catch (e: ClosedReceiveChannelException) {
                            // Do nothing!
                        } catch (e: Throwable) {
                            e.printStackTrace()
                        }
                    }
                }
    
                handleWebSocketConversation("/echo") { incoming, outgoing ->
                    val textMessages = listOf("[]", "[]")
                    for (msg in textMessages) {
                        outgoing.send(Frame.Text(msg))
                        assertEquals(msg, (incoming.receive() as Frame.Text).readText())
                    }
                    assertEquals(textMessages, received)
                }
            }
        }
    }

#### Gradle

    import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

    val kotlinVersion = "1.3.0"
    val kotlinXVersion = "1.0.0"
    val junitJupiterVersion = "5.2.0"
    val ktorVersion = "1.0.0"
    val koinVersion = "1.0.1"
    val http4kVersion = "3.100.0"

    plugins {
        java
        kotlin("jvm") version "1.3.0"
    }

    group = "finova"
    version = "1.0-SNAPSHOT"

    repositories {
        maven { setUrl("http://dl.bintray.com/kotlin/kotlin-eap") }
        maven { setUrl("https://dl.bintray.com/kotlin/ktor") }
        maven { setUrl("https://dl.bintray.com/ekito/koin") }
        maven { setUrl("http://jcenter.bintray.com") }
        mavenCentral()
        jcenter()
    }

    dependencies {
        compile(kotlin("stdlib-jdk8"))
        compile("org.jetbrains.kotlin", "kotlin-reflect", kotlinVersion)
        compile("org.jetbrains.kotlinx", "kotlinx-coroutines-core", kotlinXVersion)
    //    compile("org.jetbrains.kotlinx", "kotlinx-serialization-runtime-common", "0.9.0")

        //Ktor
        compile("io.ktor", "ktor-client-websocket", ktorVersion)
        compile("io.ktor", "ktor-client-core", ktorVersion)
        compile("io.ktor", "ktor-client-cio", ktorVersion)
    //    compile("io.ktor", "ktor-client-json", ktorVersion)
        compile("io.ktor", "ktor-websockets", ktorVersion)
        compile("io.ktor", "ktor-server-netty", ktorVersion)
        compile("io.ktor", "ktor-gson", ktorVersion)

        //Config
        compile("io.github.config4k", "config4k", "0.4.1")

        //Dependency Injection
        compile("org.koin", "koin-ktor", koinVersion)
        compile("org.koin", "koin-core", koinVersion)
        compile("org.koin", "koin-test", koinVersion)

        //Logging
        compile("org.slf4j", "slf4j-api", "1.7.25")
        compile("ch.qos.logback", "logback-classic", "1.2.3")
        compile("ch.qos.logback", "logback-core", "1.2.3")

        compile("com.google.code.gson", "gson", "2.8.5")

        // http
        compile("org.apache.httpcomponents", "httpclient", "4.5.6")
        compile("org.http4k", "http4k-client-websocket", http4kVersion)
        compile("org.http4k", "http4k-client-apache-async", http4kVersion)

        //Test
        testCompile("io.ktor", "ktor-server-test-host", ktorVersion)
        testCompile("org.junit.jupiter", "junit-jupiter-api", junitJupiterVersion)
        testImplementation("io.mockk", "mockk", "1.8.12.kotlin13")
        testRuntime("org.junit.jupiter", "junit-jupiter-engine", junitJupiterVersion)
    }

    configure<JavaPluginConvention> {
        sourceCompatibility = JavaVersion.VERSION_1_8
    }

    tasks.withType<KotlinCompile> {
        kotlinOptions.jvmTarget = "1.8"
    }

**Comment by coder82** [comment]
I just found this ticket after having lost the whole morning understanding why my tests were hanging...
I turns out the only addition I made to my app is:

```
    install(ContentNegotiation) {
        gson {}
    }
```

Then I rerun my tests and it hangs inside handleWebSocketConversation forever

**Comment by cy6erGn0m** [comment]
This bug is caused by accidentally invoked content negotiation when attempted to receive a byte channel (that is always wrong). So potentially not only websockets are affected.

Workaround is the following

```kotlin
object WorkaroundNegotiationBugConverter : ContentConverter {
    override suspend fun convertForSend(
        context: PipelineContext<Any, ApplicationCall>,
        contentType: ContentType,
        value: Any
    ): Any? {
        return null
    }

    override suspend fun convertForReceive(context: PipelineContext<ApplicationReceiveRequest, ApplicationCall>): Any? {
        return context.subject.value
    }
}

// configure content negotiation
application.install(ContentNegotiation) {
    // this should be before any content converters registrations
    register(ContentType.parse("nonexisting/mime"), WorkaroundNegotiationBugConverter)
    
    // gson { }
}
```

**Comment by cy6erGn0m** [comment]
Fixed in 1.0.1
